src/glmfit.jl

```diff
@@ -52,35 +56,76 @@ end
 
 Update the mean, working weights and working residuals, in `r` given a value of
 the linear predictor, `linPr`.
-
-This the fallback method.  Specialized methods are defined for various
-distributions and their canonical link functions.
 """
 function updateμ!{T<:FPVector,D,L}(r::GlmResp{T,D,L}, linPr::T)
-    y, η, μ, μη, wts, wrkwt, wrkres, devres =
-        r.y, r.eta, r.mu, r.mueta, r.wts, r.wrkwt, r.wrkresid, r.devresid
+    isempty(r.offset) ? copy!(r.eta, linPr) : broadcast!(+, r.eta, linPr, r.offset)
+    updateμ!(r)
+    if !isempty(r.wts)
+        r.devresid .*= r.wts
+        r.wrkwt .*= r.wts
+    end
+    r
+end
+
+function updateμ!{T<:FPVector,D<:Union{Bernoulli,Binomial}}(r::GlmResp{T,D,LogitLink})
+    y, η, μ, wrkres, wrkwt, dres = r.y, r.eta, r.mu, r.wrkresid, r.wrkwt, r.devresid
+
+    @inbounds Threads.@threads for i in eachindex(μ)
+        ηi = clamp(η[i], -20.0, 20.0)
+        ei = exp(-ηi)
+        opei = 1 + ei
+        μi = μ[i] = inv(opei)
+        dμdη = wrkwt[i] = ei / abs2(opei)
+        yi = y[i]
+        wrkres[i] = (yi - μi) / dμdη
+        dres[i] = -2 * (yi == 1 ? log(μi) : yi == 0 ? log1p(-μi) :
+            (yi * (log(μi) - log(yi)) + (1 - yi) * (log1p(-μi) - log1p(-yi))))
+    end
+end
+
+function updateμ!{T<:FPVector,D<:Poisson}(r::GlmResp{T,D,LogLink})
+    y, η, μ, wrkres, wrkwt, dres = r.y, r.eta, r.mu, r.wrkresid, r.wrkwt, r.devresid
 
-    isempty(r.offset) ? copy!(η, linPr) : broadcast!(+, η, linPr, r.offset)
-    @inbounds @simd for i = eachindex(η, μ, μη, y, wrkres, devres)
+    @inbounds Threads.@threads for i in eachindex(η)
         ηi = η[i]
+        μi = μ[i] = exp(ηi)
+        dμdη = wrkwt[i] = μi
+        yi = y[i]
+        wrkres[i] = (yi - μi) / dμdη
+        dres[i] = 2 * (xlogy(yi, yi / μi) - (yi - μi))
+    end
+end
+
+function updateμ!{T<:FPVector,D<:Normal}(r::GlmResp{T,D,IdentityLink})
+    y, η, μ, wrkres, wrkwt, dres = r.y, r.eta, r.mu, r.wrkresid, r.wrkwt, r.devresid
+
+    @inbounds Threads.@threads for i in eachindex(η)
+        μi = μ[i] = η[i]
+        wrkwt[i] = 1
+        yi = y[i]
+        wrkresi = wrkres[i] = (yi - μi)
```

Comment:
I don't think using a temporary `wrkresi` variable makes any difference in performance, especially not inside an `@inbounds` block.